### PR TITLE
Move `ImpTest` to Shelley testlib

### PR DESCRIPTION
**Pull request by Soupstraw**
# Description

Moved `ImpTest` to core and added the `EraImpTest` class together with an instance in the `Conway` era. Moved some tests into `ImpTestSpec` in Shelley and made it run in every era.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs [hunk 1]
```diff
@@ -252,7 +253,8 @@ class EraTx era => EraUTxO era where
   -- for the transaction to be valid, once those scripts are evaluated
   getScriptsNeeded :: UTxO era -> TxBody era -> ScriptsNeeded era
 
-  -- | Extract the set of all script hashes that are needed for script validation.
+  witsVKeyNeeded :: UTxO era -> TxBody era -> GenDelegs (EraCrypto era) -> Set (KeyHash 'Witness (EraCrypto era))
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: Why is this change needed? Especially how is this relevant for the work on a test suite?

I am against adding such function to this type class, unless there is a really strong reason for it. Problem with placing this function into a type class is that it suppose work for all eras, however, starting with Conway we have no notion of `GenDelegs`.

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs [hunk 2]
```diff
@@ -252,7 +253,8 @@ class EraTx era => EraUTxO era where
   -- for the transaction to be valid, once those scripts are evaluated
   getScriptsNeeded :: UTxO era -> TxBody era -> ScriptsNeeded era
 
-  -- | Extract the set of all script hashes that are needed for script validation.
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: Why was the haddock deleted?

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs [hunk 3]
```diff
@@ -56,6 +57,9 @@ newtype AlonzoScriptsNeeded era
   = AlonzoScriptsNeeded [(ScriptPurpose era, ScriptHash (EraCrypto era))]
   deriving (Semigroup, Monoid)
 
+alonzoScriptsSet :: AlonzoScriptsNeeded era -> Set.Set (ScriptHash (EraCrypto era))
+alonzoScriptsSet (AlonzoScriptsNeeded sn) = Set.fromList $ snd <$> sn
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs]: Why recreating this function and replacing with the identical implementation of `getAlonzoScriptsHashesNeeded`?

File: eras/alonzo/impl/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 ## 1.5.0.0
 
+* Add `alonzoScriptsSet`
```
Comment (lehins) [inline eras/alonzo/impl/CHANGELOG.md]: Again, `getAlonzoScriptsHashesNeeded` is already exported and available for use outside of `cardano-ledger-alonzo`
```suggestion
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs [hunk 5]
```diff
@@ -433,6 +420,7 @@ validateMissingScripts pp (ShelleyScriptsNeeded sNeeded) scriptsprovided =
         MissingScriptWitnessesUTXOW (sNeeded `Set.difference` sProvided)
   where
     sProvided = Map.keysSet $ unScriptsProvided scriptsprovided
+    sNeeded = getScriptsHashesNeeded sn
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs]: That's a nice improvement of consistency. :+1:

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+impRootTxIdL :: Lens' (ImpTestState era) (TxId (EraCrypto era))
+impRootTxIdL = lens impRootTxId (\x y -> x {impRootTxId = y})
+
+class
+  ( Show (NewEpochState era)
+  , ToExpr (NewEpochState era)
+  , Signal (EraRule "LEDGER" era) ~ Tx era
+  , BaseM (EraRule "LEDGER" era) ~ ReaderT Globals Identity
+  , STS (EraRule "LEDGER" era)
+  , Era era
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: implied by `EraGov` and `EraUTxO`
```suggestion
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Utils.hs [hunk 7]
```diff
@@ -1,9 +1,28 @@
+{-# LANGUAGE RankNTypes #-}
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Utils.hs]: It's not used in this module 
```suggestion
```

File: libs/cardano-ledger-core/CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 ## 1.8.0.0
 
+* Add `witsVKeyNeeded` to `EraTx`
```
Comment (lehins) [inline libs/cardano-ledger-core/CHANGELOG.md]: This function should not be part of `EraUTxO` (nor should it be part of `EraTx` :smile:)
```suggestion
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+    impIOMsg "Failed to run TICK" $
+      pure $
+        runShelleyBase (applySTSTest @(EraRule "TICK" era) trc)
+  case res of
+    Right !x -> do
+      impLastTickL += 1
+      impNESL .= x
+    Left e ->
+      impIO . error $
+        "Failed to run TICK:\n" <> unlines (show <$> e)
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]:
```suggestion
        unlines $ "Failed to run TICK:" : map show e
```

**Review by lehins — CHANGES_REQUESTED** [review]
Overall this looks good. I don't really like reshuffling of `witsVKeyNeeded`.

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs (hunk 1 above)
Comment (Soupstraw) [reply libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: I used it to construct the set of witnesses for a transaction in the `mkTxWits` function. I also noticed that we had copied the whole `Utxow` rule just to swap out the `witsVKeyNeeded` function in Conway. I figured that it makes sense to have this as a class method anyways so we can better reuse code across eras.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 10]
```diff
@@ -1,682 +1,375 @@
...
+
+-- | Logs the ratios of accepted votes per category
+logAcceptedRatio :: ConwayEraGov era => GovActionId (EraCrypto era) -> ImpTestM era ()
+logAcceptedRatio aId = do
+  dRepRatio <- calculateDRepAcceptedRatio aId
+  committeeRatio <- calculateCommitteeAcceptedRatio aId
+  logEntry "----- ACCEPTED RATIOS -----"
+  logEntry $ "DRep accepted ratio:\t\t" <> show dRepRatio
+  logEntry $ "Committee accepted ratio:\t" <> show committeeRatio
+  logEntry ""
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:347]: Why not add a function that can be applied to lists?
```suggestion
  logEntries 
    [ "----- ACCEPTED RATIOS -----"
    , "DRep accepted ratio:\t\t" <> show dRepRatio
    , "Committee accepted ratio:\t" <> show committeeRatio
    ]
```

Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:347]: Or even better takes a special `header` argument and adds the dashes around it:
```haskell
logEntries header xs = ...
```
would result in: 
```
----- Header ------
```

File: libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs (hunk 1 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs]: As we discussed in a call, this is far from being the only reason why the Utxow was recreated from scratch in almost every era. There are major differences in all of them.

Since this functionality is needed in tests it should be solved in tests. So, adding such function to a type class in a test suite is a totally sensible solution.

Also a request for the future. Could you please separate any serious changes like that, which affect the rules, into a separate smaller PR. Mixing restructures like that to the rules with refactoring of test suites could obfuscate potential bugs.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs (hunk 10 above)
Comment (Soupstraw) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:347]: I think an even better solution would be to format logs with the prettyprinter library.

Comment (lehins) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:347]: Sure, that could be. We'll solve this problem at a different time. This was just a mere suggestion.

Comment (Soupstraw) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:347]: I switched the logging type to `Doc ()`, this should add more flexibility to how we want to format the logs in the future.

**Review by lehins — APPROVED** [review]
Beautiful! Thank you!
